File: engine/nexgen_engine/sanity/checks/prompt_quality.py
"""Per-shot visual-prompt quality (length + generic-slop heuristics).

- PROMPT_TOO_SHORT (error): a prompt this short cannot carry the required
  components (subject+action / position / setting / camera / light+mood).
- PROMPT_THIN (warn): borderline — one of those components is probably too terse.
- PROMPT_GENERIC (warn): generic adjectives ("epic", "cinematic masterpiece")
  without concrete image description — slop risk.

This is the format-neutral core of the music pack's richer `prompts` check; the
metaphor / undefined-group / title-card / blocking heuristics stay in the pack
because they depend on pack-specific validators.
"""

from __future__ import annotations

from nexgen_engine.sanity.audit import AuditContext
from nexgen_engine.sanity.models import Finding

_SHORT_LEN = 60
_THIN_LEN = 120
_GENERIC_TOKENS = ("epic", "cinematic masterpiece")
_GENERIC_MAX_LEN = 200


def check(ctx: AuditContext) -> list[Finding]:
    out: list[Finding] = []
    for shot in ctx.shotlist.shots:
        p = shot.visual_prompt.strip()
        if len(p) < _SHORT_LEN:
            out.append(
                Finding(
                    "error", "PROMPT_TOO_SHORT", shot.id,
                    f"visual_prompt only {len(p)} chars — missing required "
                    "components (subject+action / position / setting / "
                    "camera / light+mood)",
                )
            )
        elif len(p) < _THIN_LEN:
            out.append(
                Finding(
                    "warn", "PROMPT_THIN", shot.id,
                    f"visual_prompt only {len(p)} chars — one of the required "
                    "components is probably too terse",
                )
            )
        lower = p.lower()
        if any(w in lower for w in _GENERIC_TOKENS) and len(p) < _GENERIC_MAX_LEN:
            out.append(
                Finding(
                    "warn", "PROMPT_GENERIC", shot.id,
                    "generic adjectives without concrete image description — slop risk",
                )
            )
    return out

File: Sources/NexGenEngine/Sanity/Checks/PromptQuality.swift
import Foundation

/// Per-shot visual-prompt quality (length + generic-slop heuristics).
///
/// - `PROMPT_TOO_SHORT` (error): a prompt this short cannot carry the required
///   components (subject+action / position / setting / camera / light+mood).
/// - `PROMPT_THIN` (warn): borderline — one of those components is probably
///   too terse.
/// - `PROMPT_GENERIC` (warn): generic adjectives ("epic", "cinematic
///   masterpiece") without concrete image description — slop risk.
///
/// This is the format-neutral core of a pack's richer prompt checks; the
/// metaphor / undefined-group / title-card / blocking heuristics stay out of
/// the engine because they depend on pack-specific validators. Port of
/// `sanity/checks/prompt_quality.py::check`.
public let promptQualityCheck: SanityCheck = { ctx in
    let shortLen = 60
    let thinLen = 120
    let genericTokens = ["epic", "cinematic masterpiece"]
    let genericMaxLen = 200

    var out: [Finding] = []
    for shot in ctx.shotlist.shots {
        let p = shot.visualPrompt.trimmingCharacters(in: .whitespacesAndNewlines)
        if p.count < shortLen {
            out.append(
                Finding(
                    level: .error,
                    code: "PROMPT_TOO_SHORT",
                    shotId: shot.id,
                    message:
                        "visual_prompt only \(p.count) chars — missing required components "
                        + "(subject+action / position / setting / camera / light+mood)"
                )
            )
        } else if p.count < thinLen {
            out.append(
                Finding(
                    level: .warn,
                    code: "PROMPT_THIN",
                    shotId: shot.id,
                    message:
                        "visual_prompt only \(p.count) chars — one of the required components "
                        + "is probably too terse"
                )
            )
        }

        let lower = p.lowercased()
        if genericTokens.contains(where: { lower.contains($0) }) && p.count < genericMaxLen {
            out.append(
                Finding(
                    level: .warn,
                    code: "PROMPT_GENERIC",
                    shotId: shot.id,
                    message: "generic adjectives without concrete image description — slop risk"
                )
            )
        }
    }
    return out
}
